x/ecocredit/server/core/bridge_receive.go

```diff
@@ -0,0 +1,138 @@
+)
+
+// BridgeReceive bridges credits received from another chain.
+func (k Keeper) BridgeReceive(ctx context.Context, req *core.MsgBridgeReceive) (*core.MsgBridgeReceiveResponse, error) {
+	bridgeServiceAddr, err := sdk.AccAddressFromBech32(req.Issuer)
+	if err != nil {
+		return nil, err
+	}
+
+	class, err := k.stateStore.ClassTable().GetById(ctx, req.Project.ClassId)
+	if err != nil {
+		return nil, sdkerrors.ErrInvalidRequest.Wrapf("could not get class with id %s: %s", req.Project.ClassId, err.Error())
+	}
+
+	// first we check if there is an existing project
+	idx := api.ProjectClassKeyReferenceIdIndexKey{}.WithClassKeyReferenceId(class.Key, req.Project.ReferenceId)
+	it, err := k.stateStore.ProjectTable().List(ctx, idx)
+	if err != nil {
+		return nil, err
+	}
+
+	// we only want the first project that matches the reference ID, so we do not loop here.
+	var project *api.Project
+	if it.Next() {
+		var err error
+		project, err = it.Value()
+		if err != nil {
+			return nil, err
+		}
+	}
+	it.Close()
+
+	// if no project was found, create one + issue batch
+	if project == nil {
+		projectRes, err := k.CreateProject(ctx, &core.MsgCreateProject{
+			Admin:        req.Issuer,
+			ClassId:      req.Project.ClassId,
+			Metadata:     req.Project.Metadata,
+			Jurisdiction: req.Project.Jurisdiction,
+			ReferenceId:  req.Project.ReferenceId,
+		})
+		if err != nil {
+			return nil, err
+		}
+		batchRes, err := k.CreateBatch(ctx, &core.MsgCreateBatch{
+			Issuer:    req.Issuer,
+			ProjectId: projectRes.ProjectId,
+			Issuance: []*core.BatchIssuance{
+				{Recipient: req.Batch.Recipient, TradableAmount: req.Batch.Amount},
+			},
+			Metadata:  req.Batch.Metadata,
+			StartDate: req.Batch.StartDate,
+			EndDate:   req.Batch.EndDate,
+			Open:      true,
+			OriginTx:  req.Batch.OriginTx,
+			Note:      req.Batch.Note,
+		})
+		if err != nil {
+			return nil, err
+		}
+		return &core.MsgBridgeReceiveResponse{BatchDenom: batchRes.BatchDenom, ProjectId: projectRes.ProjectId}, nil
+	}
+
+	// batches are matched on their denom, iterating over all batches within the <ProjectId>-<StartDate>-<EndDate> range.
+	// any batches in that iterator that have matching metadata, are added to the slice.
+	// idx will be of form C01-001-20210107-20210125-" catching all batches with that project Id and in the date range.
+	batchIdx := fmt.Sprintf("%s-%s-%s-", project.Id, req.Batch.StartDate.Format("20060102"), req.Batch.EndDate.Format("20060102"))
+	bIt, err := k.stateStore.BatchTable().List(ctx, api.BatchDenomIndexKey{}.WithDenom(batchIdx))
+	if err != nil {
+		return nil, err
+	}
+
+	batches := make([]*api.Batch, 0)
+	for bIt.Next() {
+		batch, err := bIt.Value()
+		if err != nil {
+			return nil, err
+		}
+		// the timestamp stored in the batch is more granular than the date in the denom representation, so we match here.
+		if batch.StartDate.AsTime().Equal(*req.Batch.StartDate) &&
+			batch.EndDate.AsTime().Equal(*req.Batch.EndDate) &&
```

Comment:
Thinking back to the recent audit pull request, and noticing local time is being used in the tests above, I think we may have an issue here if the request is local time and the time stored in state is UTC...

Reply:
We should probable use `req.Batch.StartDate.UTC()` and `req.Batch.EndDate.UTC()` and we may want to test both local and UTC times in the request and always UTC time in state.

Reply:
We should probably also double check that we want to always store UTC time and we may want to open a follow up issue to make sure we have dates stored as expected and maybe update proto comments to note that the time will be stored in UTC... cc @clevinson

Reply:
switched everything to convert to UTC, also in this commit https://github.com/regen-network/regen-ledger/pull/1174/commits/794e070947ce135fa8739adeb7d82b84abd24f74

Reply:
Hmm I'm not sure I fully understand the nuance here. But we should be making sure that if a user is setting "Jan 1, 2020" that its not doing some weird local timestamp, that when converted to UTC actually reads it as "Jan 2" or "Dec 31"...  Is this something we can really ensure against?

If our intention is to only allow dates, maybe `ValidateBasic()` could ensure that the when evaluated in UTC, the timestamp % day == 00:00:00, aka that the timestamp is rounded to the nearest day @ midnight UTC?

Reply:
hmm i feel like the rounding might be okay, as long as the entire chain speaks in UTC terms? i think it would make the most sense to either drop support for explicit time values (only dates), or just speak UTC throughout the chain.

Reply:
@clevinson @technicallyty the Toucan data suffers from this problem. All the date/times are stored as UNIX date/times - e.g. without timezone data. They've obviously been entered as an exact date, but the local timezone where they've been inserted is a few hours off from UTC, so all the dates/times in the Toucan data are a few hours before the exact date. ([See raw Toucan data here on the Vintage tab](https://docs.google.com/spreadsheets/d/1NhhGPrSZoc3_ptcHpUjWn_XR71VFjuTQhu7Gf21oyk8/edit#gid=561399407)). 

We [discussed this in the meta data conversion ticket here](https://github.com/rnd-inc/toucan-bridge-service/issues/48), and decided to round the Toucan date/time data to the nearest whole day. [The conversion code is here](https://github.com/rnd-inc/toucan-bridge-service/blob/main/metadata/metadata.go#L66-L69), [here](https://github.com/rnd-inc/toucan-bridge-service/blob/main/metadata/metadata.go#L46-L47), and [here](https://github.com/rnd-inc/toucan-bridge-service/blob/main/metadata/metadata_test.go#L9-L30).

Reply:
@clevinson as per slack comment, I've just realised that the Toucan data was actually /not/ off by one hour, and that issue and all the discussion about it was nonsense. However, this highlights the importance of using `.UTC()` everywhere in Go, or you'll get times converted to whatever the local timezone of the server is!

Reply:
... however, the regen ledger code that generates the batch denom [isn't using .UTC()](https://github.com/regen-network/regen-ledger/blob/852fb8d82ecef8823f322894c8f0ce7610e31a0a/x/ecocredit/core/utils.go#L87-L91), and I assume the reason that this works fine is that the timezone of the server (and probably all servers) is set to UTC... Whereas when I ran the script to read the live data from the Toucan blockchain, I ran it on my Mac with a local  timezone set. 🤔